### PR TITLE
[Feature] Trackable Race Conditions

### PR DESCRIPTION
**Pull request by jonnii**

### CODE REVIEW

**Comment by jonnii** [comment]
@bieri i think we can get away with this. thoughts?

**Comment by jonnii** [comment]
@jonrad can you eyeball this one too?

**Comment by bieri** [comment]
@jonnii  this works and does prevent the dirty write case, but wont this end up being slower than ConcurrentDictionary<> in read heavy cases?

**Comment by jonnii** [comment]
i don't think there are any read heavy cases.

File: src/Chinchilla/TrackableFactory.cs [hunk 1]
```diff
@@ -7,27 +7,45 @@ namespace Chinchilla
...
         private readonly List<TTrackable> trackables = new List<TTrackable>();
 
         public IEnumerable<TTrackable> Tracked
         {
-            get { return trackables; }
+            get
+            {
+                lock (trackLock)
+                {
+                    return trackables.ToArray();
```
Comment (jonrad) [inline src/Chinchilla/TrackableFactory.cs]: Your Dispose does a ToArray which is now redundant

**Comment by jonnii** [comment]
FYI, the `IsTracking` method is only used in specs, and the overhead of creating/tearing down trackables (e.g. subscriptions) probably outweighs the lock contention you may see. i think.

**Comment by jonrad** [comment]
Yeah, from my understanding this would work. Obviously, adds some performance overhead which is the more important question here I believe.

**Comment by bieri** [comment]
running through this on my side i see ~780 / s  which should be more than sufficient for a single request response pair. the old code could only get me 30 - 50 /s before it would fall over.

**Comment by bieri** [comment]
@jonnii  I'm good

**Pull request merged**